### PR TITLE
Modernize discrete factor graphs (a bit)

### PR DESCRIPTION
**Pull request by dellaert**
- use emplace_shared were we can
- bring use of Values up to date with rest of GTSAM (no shared pointer mania)

Technically an API change, but discrete is very seldomly used, if ever, except by me I think.

### CODE REVIEW

File: gtsam_unstable/discrete/tests/testCSP.cpp [hunk 1]
```diff
@@ -120,8 +120,8 @@ TEST_UNSAFE(CSP, WesternUS) {
   Ordering ordering;
   ordering += Key(0), Key(1), Key(2), Key(3), Key(4), Key(5), Key(6), Key(7),
       Key(8), Key(9), Key(10);
-  CSP::sharedValues mpe = csp.optimalAssignment(ordering);
-  // GTSAM_PRINT(*mpe);
+  auto mpe = csp.optimalAssignment(ordering);
+  // GTSAM_PRINT(mpe);
   CSP::Values expected;
   insert(expected)(WA.first, 1)(CA.first, 1)(NV.first, 3)(OR.first, 0)(
```
Comment (ProfFan) [inline gtsam_unstable/discrete/tests/testCSP.cpp:126]: Sorry for jumping in, but I remember last time you asked me to wrap some discrete in Python, and I realized this is not easy to wrap. Could we drop this syntax (which is boost specific) and use some C++11-only style instead?

Comment (dellaert) [reply gtsam_unstable/discrete/tests/testCSP.cpp:126]: different PR

**Comment by xsj01** [comment]
Looks good to me

**Review by xsj01 — APPROVED** [review]

**Comment by dellaert** [comment]
Thanks @xsj01, but note reviewers don't merge :-)

**Pull request merged**